### PR TITLE
New dark theme and glyph icons #4 #50

### PR DESCRIPTION
**Pull request by 0xaryan**
### Editor

- Editor currently has a new theme. It's fully dark based on Dark Blend colors #50 
- Window Chrome icons and Xenko logo replaced with glyph versions
- Chrome is now a little smaller
- Support for `HslColor` and conversation methods to Rgb (`System.Color`)

Dark Editor

### Menus
- Menus redesigned. Theirs margins/paddings are changed.
- Menus are now fully dark. (#101011)

Dark Menu

### Icons
- A huge number of small png icons and almost all of the icons in the Image Dictionaries replaced with flat responsive icons of visual studio. #4 
- Icons now support Luminosity Transform. It means, they can automatically change color based on the theme of the editor. Currently all the transformation are set to dark.
I believe that this mechanism can be used for future features of Theming (Support for Light theme as an example)
- New `ImageThemingUtilities` class API that changes the luminosity of drawing image.
- New `ThemedSource` markup extension that can change source of an icon to another theme on the fly.
`<Image Source="{xk:ThemedSource {StaticResource ImageOpenFolder}, Dark}" />`

Icons

### Job is not done.
It is obvious that this new skin is not still good enough, but for me, as a student and a newbie to UI and game designing it took only a few days to understand the project and make these changes.
So to the great and professional community of Xenko, please make this engine even better.

Xenko

### CODE REVIEW

**Comment by ahrenm** [comment]
Ohhhh can haz plz :-)

**Comment by xen2** [comment]
Thanks for this nice contribution!
Just to make sure, you only used icons from the Visual Studio Image Library, nothing else?

**Comment by 0xaryan** [comment]
@RUSshy 

I'm totally agree with you. As I said in the comment, the editor is not ideal nor consistent yet.
My time is really limited to give Xenko editor a fully flat look and feel; purpose of these changes were only a try making my favorite game engine a little bit better, or at least to give others some inspirations for further work.

I think your suggestions are near to Blend design guidelines, as it uses colored icons for menus, solution explorer,... and monochromatic icon for the most parts of the editor and designer. Besides of that, I believe even if we use colored icons, if we choose right colors in the right layouts, we still have a nice looking editor (just like Godot has).

**Comment by Kryptos-FR** [comment]
I just noticed that @0xaryan merged master back into the branch. @xen2 will that cause an issue to keep a linear history, as you were doing until now? Or do you rebase manually?

If so we should maybe add a rule that merging master should be avoided when possible, unless there are some conflicts to resolve.

**Comment by xen2** [comment]
@Kryptos-FR it's not a problem to merge master, it is often necessary to do so to make the branch mergeable back later. I can still squash+rebase using github UI.

**Comment by Kryptos-FR** [comment]
@xen2 I have a few fixes to add. How can I push new commits to this PR?

@0xaryan If possible try to not needlessly reformat files. The changes in `Theme.xaml` are very hard to review because of a lot of spacing and attribute order have changed.

**Comment by xen2** [comment]
@Kryptos-FR I see two possibilities:
1. Your fixes are just before the merge, you can add on top of the PR and maybe open a new PR with it
2. Your fixes are part of the PR but further changes are needed: submit a PR to https://github.com/0xaryan/xenko (anything in its `master` (the PR source branch) will be part of this PR)

**Comment by Kryptos-FR** [comment]
Let's try option 2.

**Comment by 0xaryan** [comment]
@Kryptos-FR 

> If possible try to not needlessly reformat files. The changes in Theme.xaml are very hard to review because of a lot of spacing and attribute order have changed.

I have a XAML beautifier extension installed on vs that make those changes. I turned it off from the middle of the work.

**Comment by xen2** [comment]
Forgot the obvious option to work in a repo where both of you have writing permissions on that feature branch.

**Comment by mafiesto4** [comment]
Good luck with updating documentation screenshots! 😄

File: sources/presentation/Xenko.Core.Presentation/Themes/ExpressionDark/Theme.xaml [hunk 1]
```diff
@@ -2299,39 +2302,56 @@
...
+                    <ColumnDefinition Width="*" />
+                    <ColumnDefinition Width="1" />
+                  </Grid.ColumnDefinitions>
+                  <Border Grid.Row="0" Grid.ColumnSpan="3" Background="{StaticResource MenuBackgroundBrush}" />
+                  <Border Grid.Row="2" Grid.ColumnSpan="3" Background="{StaticResource MenuBackgroundBrush}" />
+                  <Border x:Name="SubmenuMainWrapperBorder" Grid.RowSpan="3" Grid.ColumnSpan="3" Background="{StaticResource MenuBackgroundBrush}" BorderBrush="{StaticResource MenuBorderBrush}" BorderThickness="0.5" />
+                  <Grid x:Name="SubMenu" Grid.Row="1" Grid.Column="1">
+                    <Border x:Name="SubMenuBorder" Background="{DynamicResource MenuBackgroundBrush}" BorderBrush="{DynamicResource MenuBorderBrush}" BorderThickness="0" />
+                    <StackPanel IsItemsHost="True" KeyboardNavigation.DirectionalNavigation="Cycle" Grid.IsSharedSizeScope="true" />
+                  </Grid>
```
Comment (Kryptos-FR) [inline sources/presentation/Xenko.Core.Presentation/Themes/ExpressionDark/Theme.xaml]: Something here is breaking the "view options" menu in the asset view: text and icon are incorrectly aligned (horizontally). But I can't figure exactly where the issue is: Margin, Border, Padding or a combination or something else (the issue could be a side-effect with the style of that menu in the AssetViewUserControl.xaml).

image
image


What was the reason for changing the submenu template here?

File: sources/editor/Xenko.Core.Assets.Editor/View/DefaultPropertyTemplateProviders.xaml [hunk 2]
```diff
@@ -133,7 +133,7 @@
     <Setter Property="ContentTemplate">
       <Setter.Value>
         <DataTemplate>
-          <Image Source="{StaticResource ImageRemove}" Width="16" Height="16" RenderOptions.BitmapScalingMode="NearestNeighbor"/>
+          <Image Source="{DynamicResource ImageClose}" Width="16" Height="16" RenderOptions.BitmapScalingMode="NearestNeighbor"/>
```
[the remark this answers is not in the document]
Comment (0xaryan) [reply sources/editor/Xenko.Core.Assets.Editor/View/DefaultPropertyTemplateProviders.xaml]: `{DynamicResource}` is not needed here. I changed it for some markup extension combining test in my old changes, but I forgot to change it back to `{StaticResource ImageRemove}`.

File: sources/presentation/Xenko.Core.Presentation/Themes/ExpressionDark/Theme.xaml (hunk 1 above)
Comment (0xaryan) [reply sources/presentation/Xenko.Core.Presentation/Themes/ExpressionDark/Theme.xaml]: Humm... I'll take a look.

Comment (0xaryan) [reply sources/presentation/Xenko.Core.Presentation/Themes/ExpressionDark/Theme.xaml]: Weird but, it looks fine in my side 🤔 

View Options

Comment (Kryptos-FR) [reply sources/presentation/Xenko.Core.Presentation/Themes/ExpressionDark/Theme.xaml]: Hmm. Really weird. I cleaned the whole solution and recompiled and now the issue is also gone on my side.

Maybe some XAML files were not recompiled correctly and it was mixing the new styles with some remains of the old one.

Comment (0xaryan) [reply sources/presentation/Xenko.Core.Presentation/Themes/ExpressionDark/Theme.xaml]: @Kryptos-FR I saw a lot of weird behaviors from visual studio when I was working on this project, specially when working with Edit-and-Continue. Sometimes I need to clean whole solution even for a small change to take effect.

**Comment by RUSshy** [comment]
I think it could be pushed into a different branch, so we can also provide our contributions so once project finish, we'll be all ready to update documentation screenshot before merging into master as @mafiesto4 said

**Comment by Timurid** [comment]
Freaking. Good.

**Comment by 0xaryan** [comment]
@Kryptos-FR Alright (:

**Comment by 0xaryan** [comment]
Beside the previous changes, new changes made to Scene Tabs, Menus, Property Grid title, Footer status bar and some color fixings. I'm not going to make any other changes in this PR, but only fixing review issues.
_But there are still changes that can be made (specially in AvalonDock theme) by the community in further developments._

**Comment by Kryptos-FR** [comment]
@0xaryan Sure. I think it is good enough at the moment. We can always improve later.

Thanks for the contribution. I'll do another quick review and then I think we can merge.

**Comment by 0xaryan** [comment]
I talked with one of the [Icons8](https://icons8.com) supports. They said:
> Actually all logos should be free, but it seems that this icon is in other category, that's why it's not shown as free . I will surely pass it to our team to look into! Can you please tell me which size do you need this icon in - I will send them to you . And if you need other logos - sen me the links and sizes, I will send them as well 

They sent icons I've requested (Apple, Windows, UWP, Android, Linux and iOS) in SVG format, so we can replace OS icons too.
They have a specific license described [here](https://icons8.com/license/). 
All it needs is to set a link to Icons8 in the about dialog of Xenko. Is there any problem about this?

**Comment by Kryptos-FR** [comment]
> All it needs is to set a link to Icons8 in the about dialog of Xenko. Is there any problem with this?

We have links to the libraries we use in the [THIRD PARTY.md](https://github.com/xenko3d/xenko/blob/master/THIRD%20PARTY.md) document. The document is available here and also from the about dialog through a link. I suppose it complies to their license if we add it there (even though it is one click away).

**Comment by 0xaryan** [comment]
@Kryptos-FR Ok, I'll change those icons.

**Comment by Toriadus** [comment]
Variant Dark Xenko isn't the most successful from the point of visual perception , too sharp contrast of black color and light colors. My opinion - the ideal option would be a gray color with a light gradient of monochrome icons like Blender 3D, Modo, Atomic Game Engine, Wave Game Engine, Shiva Game Engine, etc. This is the most favorable option for the eyes.

blender264_interface
robobeach2017
sffbzqr
mosque3
makinggameassetsw_modo_scrnsht02

**Comment by Toriadus** [comment]
Here is an example of the interface in Amazon Lumberyard. Very well readable and pleasing to the eyes light gray icons with a light gray gradient on the background of a slightly darker gray interface. The interface as you can see consists of two variants of gray color, more dark gray and slightly lighter, this combination looks very nice.
amazon-lumberyard

**Comment by 0xaryan** [comment]
@Toriadus Thanks for sharing.

> the ideal option would be a gray color with a light gradient of monochrome icons.

I'm not fan of a gray color editor, and I think most people prefer to have a light or dark theme, not something in the middle. Just take a look at the most voted request of Unity3d community feedback, near 13000 votes for a free dark theme.
To me it was the best option to use Dark UI of Blend. As Xenko is a game engine built on .Net and it's API is the most comfortable API for .Net developers, why not an editor like Visual Studio?

Blend

So, tried to transform the gray editor of Xenko from this:
Gray Xenko

To this:
Dark Xenko

I barely saw that someone change clean theme of Visual Studio to something fully different.
But all these ideas and styles that some people like and some people don't can be fixed when a theming system comes to the Xenko.

**Comment by Toriadus** [comment]
Thank you for your feedback, my previous reply is my subjective opinion. I got acquainted  with your version of the new theme more closer and I can say that your option is much better than the previous one. Color icons look more attractive (without a gradient as before), new icons of light gray color (movement, camera, is /) look more interesting, thanks for the work done, but I still have questions (my opinion).
I got acquainted with the color gamut of the interface of the top ten most popular game engines and none of them didn`t  have so much dark color, even the dark theme of Godot is much lighter than this option. 

You said that most people prefer to have a light or dark theme, but in fact I gave the example of the popular engines and none of them doesn't own as much a dark color theme as in your version. Once again I repeat that in your version of the dark theme the color gamut is too dark (approaching black) and against the background of light inscriptions creates a very sharp and uncomfortable contrast for the eyes (this is my opinion), I have not seen anything like this in any of the engines. Now a few words about my recommendations and advice:

- As someone said, icons would not hurt to be grouped by purpose
- The colored icons that you left, you can also make a light gray
- It also seems to me that the windows don`t have lack a small chamfer with a glare on the edges, so the menu bars will be seem to stand out from the rest, it can be well traced in the UI amazon-lumberyard

Thanks again for your work, I appreciate it.

amazon-lumberyard21

**Comment by 0xaryan** [comment]
@Toriadus 
> You said that most people prefer to have a light or dark theme, but in fact I gave the example of the popular engines and none of them doesn't own as much a dark color theme as in your version. 

I didn't mean that the examples you gave are not dark, except the Blender and MODO that are semi-dark, all others are dark. But this is what I'm talking about:

1. [Unity Feedback Editor: Dark skin theme in Free available in options.](https://feedback.unity3d.com/forums/unity/suggestions?utf8=%E2%9C%93&status=0&category=&view=most_voted)
2. [How can I change Blender's theme?](https://blender.stackexchange.com/questions/3442/how-can-i-change-blenders-theme)

In Blender, you can easily change the theme in options. Maybe that's why they are not worried about changing default theme to a darker one.
In Unity, I think it is a part of their business model. Dark theme is only available in the professional edition and this became a revenue stream for unity developers.

As I said before, these are all preferences. I prefer to have a dark UI just like I have in Blend editor, and I know that there are lots of folks out there that agree with me and lots of people who don't.
The only thing that can solve these type of style issues, is to bring more themes into the editor to support more people.

Supporting more theme is not something that can happen anytime soon as it needs a lot of changes and active work, so the best option is to change the editor to something that the most members of the community like.
When I was creating this new theme, I was afraid that the community don't accept it as it is very dark (as you said), so I asked some people as a small statistical people about this new color and they all said that it is OK.

Besides your helpful advices, I have other ideas that could make Xenko editor much cleaner and better, but my time is very limited to implement them.

Anyway, I do not insist on this color theme. If you think that a lighter color is a better choice for Xenko, you are only a PR away.

> Thanks again for your work, I appreciate it.

Your welcome @Toriadus (: 🙏

**Comment by Toriadus** [comment]
Thanks for the answer, adding themes to options is really a good idea! Is it possible to general overview several options for an UI with different colors gamma and different configurations of the layout of icons? In any case, I respect the opinion of the community and will always listen to it.

**Comment by xen2** [comment]
> Besides your helpful advices, I have other ideas that could make Xenko editor much cleaner and better, but my time is very limited to implement them.
@0xaryan If you don't have the time, feel free to open an issue and if it's good, somebody else might take care of it!

**Comment by xen2** [comment]
Thanks!

**Pull request merged**